osu.Game/Beatmaps/Drawables/StarRatingDisplay.cs

```diff
@@ -0,0 +1,100 @@
+using osu.Framework.Bindables;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.UserInterface;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Overlays;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Beatmaps.Drawables
+{
+    /// <summary>
+    /// A pill that displays the star rating of a beatmap.
+    /// </summary>
+    public class StarRatingDisplay : CompositeDrawable, IHasCurrentValue<StarDifficulty>
+    {
+        private readonly Box background;
+        private readonly SpriteIcon starIcon;
+        private readonly OsuSpriteText starsText;
+
+        private readonly BindableWithCurrent<StarDifficulty> current = new BindableWithCurrent<StarDifficulty>();
+
+        public Bindable<StarDifficulty> Current
+        {
+            get => current.Current;
+            set => current.Current = value;
+        }
+
+        [Resolved]
+        private OsuColour colours { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private OverlayColourProvider colourProvider { get; set; }
+
+        /// <summary>
+        /// Creates a new <see cref="StarRatingDisplay"/> using an already computed <see cref="StarDifficulty"/>.
+        /// </summary>
+        /// <param name="starDifficulty">The already computed <see cref="StarDifficulty"/> to display.</param>
+        public StarRatingDisplay(StarDifficulty starDifficulty)
+        {
+            Current.Value = starDifficulty;
+
+            Size = new Vector2(52f, 20f);
+
+            InternalChild = new CircularContainer
+            {
+                Masking = true,
+                RelativeSizeAxes = Axes.Both,
+                Children = new Drawable[]
+                {
+                    background = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                    },
+                    starIcon = new SpriteIcon
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Margin = new MarginPadding { Right = 30f },
+                        Icon = FontAwesome.Solid.Star,
+                        Size = new Vector2(8f),
+                    },
+                    starsText = new OsuSpriteText
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Margin = new MarginPadding { Left = 10f, Bottom = 2f },
+                        // todo: this should be size: 12f, but to match up with the design, it needs to be 14.4f
+                        // see https://github.com/ppy/osu-framework/issues/3271.
+                        Font = OsuFont.Torus.With(size: 14.4f, weight: FontWeight.Bold),
+                        Shadow = false,
+                    }
```

Comment:
Quite confounded by this layout logic. It feels very strange and brittle (centre-anchored text with weird margins?). In fact it breaks already with some rather extreme maps.

![2021-08-07-160620_91x67_scrot](https://user-images.githubusercontent.com/20418176/128602961-e72da0cd-d3f4-4313-bb41-6ba3e1b33840.png)

I really do think that this should use a grid at least, with two columns: one for the stars, and one auto-size min-width column for the centered text.

Reply:
I'll give that a shot, min-width auto-sizing sounds like it would work exactly the way I hoped it to be, good idea on that one.

Reply:
In the screenshot, both should be left aligned if anything.

The padding inside the pill in that screenshot also looks wrong (not enough on the right hand side).

Reply:
I've refactored this into a grid container with minimum size, no longer percisely matching the design but looks pretty good and still visually almost match.

- regular size:
![CleanShot 2021-08-18 at 10 57 49](https://user-images.githubusercontent.com/22781491/129860822-ebbde980-8adf-4913-8b31-9d02900e9271.png)

- small size (used in beatmap cards and what not):
![CleanShot 2021-08-18 at 10 57 55](https://user-images.githubusercontent.com/22781491/129860827-66454179-6960-4c0a-a25e-4c4d282be191.png)

Reply:
I suppose that looks better, although the padding on the right side of the text changes depending on its width (especially apparent on the ones ending with `.10` above).

Reply:
You raised a good point, it looks like the design make the text centered rather than centre-left, and along with the flexibility of the display, the displays look yet better now:
![CleanShot 2021-08-18 at 11 46 42](https://user-images.githubusercontent.com/22781491/129868075-3279f976-7bb9-4526-96d3-a290b57262dd.png)
![CleanShot 2021-08-18 at 11 46 48](https://user-images.githubusercontent.com/22781491/129868061-a1f1e739-1f7c-46dc-841d-9d5bea2519a7.png)